### ISSUE TITLE
Quantity initializer bug?

### ISSUE THREAD

**Issue by thomaslima** [hgrecco/pint#750]
I don't know if this is intended behavior or not, but it seems by your `__new__` advanced definition that you wanted to cover all cases. So here's a test that I think fails in my opinion.

```python
Python 3.6.7 (default, Jan  2 2019, 14:51:42)
[GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)] on darwin
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry()
>>> Q_ = ureg.Quantity
>>> x = Q_(1, 'm')
>>> y = Q_(1000, 'mm')
>>> z = Q_(x, 'mm')
>>> x, y, z
(<Quantity(1, 'meter')>, <Quantity(1000, 'millimeter')>, <Quantity(1 meter, 'millimeter')>)
>>> x == y
True
>>> x == z
False
>>> y == z
False
```

Is there any reason to not have `x == y == z`? `<Quantity(1 meter, 'millimeter')>` looks odd to me.

I came across this when trying to create a function that could receive quantities as arguments. I am going to use the `ureg.wraps` decorator, as I believe you intended in the documentation. But my first instinct was to get the argument as x, and then initialize `Q_(x, 'mm')` so that the units were converted automatically, or created by default.

Thank you for this package!

**Comment by hgrecco** [comment]
Initialize a Quantity with another quantity and units is discouraged (maybe we should raise an error). I do not see any place in which this is useful. Shall we raise a ValueException?

**Comment by dimapu** [comment]
A nice feature related to this question could be a function `as_quantity(value, unit)`:

```python
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry()
>>> Q_ = ureg.Quantity
>>> q = Q_(1, 'm')

>>> q1 = ureg.as_quantity(1, 'm')  # can be used to create a quantity from a num
>>> q1
<Quantity(1, 'meter')>

>>> q2 = ureg.as_quantity(q1, 'm')  # can be used to create a quantity from another quantity
>>> q2
<Quantity(1, 'meter')>

>>> q3 = ureg.as_quantity(q1, 'mm')  # ignores the unit if input is a quantity
>>> q3
<Quantity(1, 'meter')>

>>> q4 = ureg.as_quantity(q1, 'A')  # but checks the dimensionality
Exception: DimentionalityError...

>>> ureg2 = UnitRegistry()
>>> qu = ureg2.Quantity(1, 'm')
>>> q5 = ureg.as_quantity(qu, 'm')  # converts to the ureg UnitRegistry
```

**Comment by jondoesntgit** [comment]
I agree with @hgrecco, I can’t think of a use case where you would want to create a quantity using another quantity. Those edge cases could be covered by `q2 = Q_(q1.m, ‘inches’)`

**Comment by dimapu** [comment]
A use case could be when treating a user input:

```python
# lib.py
class stage():
...
    def move(self, distance):
        distance = as_quantity(distance, 'um')
        self._stage_controller.move(distance.m_as('m'))
```
```python
# user_code.py
stage().move(1)  # this accepts both number
stage().move(Q_(1, 'mm'))  # and a Quantity, independent of the UnitRegistry used by a user
```

**Comment by thomaslima** [comment]
Thanks for reviewing my issue!

I am working on some python code that will do some simulations, but I don't want the end-user to be familiar with pint. 

@jondoesntgit does not work:
```python
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry()
>>> Q_ = ureg.Quantity
>>> q1 = Q_(1, 'm')
>>> q2 = Q_(q1.m, 'mm')
>>> q2
<Quantity(1, 'millimeter')>
```

Here's what I am using:
```python
# python 3.6
from pint import UnitRegistry
ureg = UnitRegistry()
Q_ = ureg.Quantity

def check_unit(quantity, unit):
    if isinstance(quantity, Q_):
        return quantity.to(unit)
    else:
        return Q_(quantity, unit)

class Simulation:
    def __init__(self, distance, speed):
        self.distance = check_unit(distance, 'm')
        self.speed = check_unit(speed, 'm/s')
```

Somehow the `wraps` decorator did not work well with classes and optional arguments. It was a hassle so I stopped using it. With the proposed functionality, I would have just written:

```python
class Simulation:
    def __init__(self, distance, speed):
        self.distance = Q_(distance, 'm')
        self.speed = Q_(speed, 'm/s')
```

In my opinion, it reads much better. In case of error, I believe whatever error `quantity.to` throws is good enough.

**Comment by jules-ch** [comment]
This use case is covered by #1282 & will be released in 0.18

Closing
